src/iterative_ensemble_smoother/_iterative_ensemble_smoother.py

```diff
@@ -88,66 +130,76 @@ def fit(
             if self.steplength_schedule is None:
-                step_length = steplength_exponential(self.iteration_nr)
+                step_length = steplength_exponential(self.iteration)
             else:
-                step_length = self.steplength_schedule(self.iteration_nr)
+                step_length = self.steplength_schedule(self.iteration)
 
         assert 0 < step_length <= 1, "Step length must be in (0, 1]"
 
-        # Seed here so we draw the same samples in every iteration (call to update())
-        rng = np.random.default_rng(self.seed)
+        if ensemble_mask is None:
+            ensemble_mask = np.ones(ensemble_size, dtype=bool)
+
+        self.ensemble_mask = ensemble_mask
 
-        # A covariance matrix was passed
-        # Columns of E should be sampled from N(0, Cdd) and centered, Evensen 2019
-        if observation_errors.ndim == 2:
-            E = rng.multivariate_normal(
-                mean=np.zeros_like(observation_values),
-                cov=observation_errors,
-                size=ensemble_size,
-                method="cholesky",  # An order of magnitude faster than 'svd'
-            ).T
-        # A vector of standard deviations was passed
-        else:
-            E = rng.normal(
-                loc=0, scale=observation_errors, size=(ensemble_size, num_obs)
-            ).T
+        # ---------------------------------------------------------------------
+        # ----------- Computations corresponding to algorithm 1 ---------------
+        # ---------------------------------------------------------------------
 
+        # Draw samples from N(0, C_dd)
+        E = self._get_E(
+            observation_errors=observation_errors,
+            observation_values=observation_values,
+            ensemble_size=ensemble_size,
+        )
         assert E.shape == (num_obs, ensemble_size)
 
-        E -= E.mean(axis=1, keepdims=True)
-
-        R, observation_errors_std = _create_errors(observation_errors, inversion)
+        # Transform covariance matrix to correlation matrix
+        R, observation_errors_std = covariance_to_correlation(observation_errors)
 
         # Store D as defined by Equation (14) in Evensen (2019)
         self.D_ = E + observation_values.reshape(num_obs, 1)
         D = self.D_ - response_ensemble
 
+        # Note on scaling
+        # -------------------
+        # In line 8 in Algorithm 1, we solve (S S^T + E E^T) X = (SW + D - g(X))
+        # for X, and then we compute S.T X. If we scale the rows (observed variables)
+        # of these equations using the standard deviations, we can obtain better
+        # conditioning numbers on the equation. This corresponds to left-multiplying
+        # with a diagonal matrix L := sqrt(diag(C_dd)). In an experiment with
+        # random covariance matrices, this improved the condition number ~90%
+        # of the time.
+        # To see the equality, note that if we scale S, E and K := (SW + D - g(X))
```

Comment:
`K` here is `H` from line 7 right? Any reason not to use `H`?

Reply:
Good catch. Should of course use H instead of introducing new variables!